### PR TITLE
Add resolving for modules and functions as function call parameters

### PR DESCRIPTION
**Pull request by telezhnaya**
Have a look, please.
I haven't written tests yet. As always, I will write them after first review. It's important for me to know your general opinion.

### CODE REVIEW

File: src/org/intellij/erlang/psi/impl/ErlangPsiImplUtil.java [hunk 1]
```diff
@@ -1821,6 +1829,48 @@ public static boolean fromTheSameCaseExpression(@NotNull PsiElement origin, @Not
...
+      super("functionCallNamedParameter");
+      NAMES = ContainerUtil.newArrayList(names);
+    }
+
+    @Override
+    public boolean accepts(@NotNull T element, ProcessingContext context) {
+      ErlangQAtom argument = element instanceof ErlangQAtom ? (ErlangQAtom) element : null;
+      if (argument == null) return false;
+
+      ErlangFunctionCallExpression call = PsiTreeUtil.getParentOfType(argument, ErlangFunctionCallExpression.class);
```
Comment (deadok22) [inline src/org/intellij/erlang/psi/impl/ErlangPsiImplUtil.java]: You're looking for function call expression recursively - it's both wrong, and bad performance-wise.

File: src/org/intellij/erlang/psi/impl/ErlangPsiImplUtil.java [hunk 2]
```diff
@@ -1821,6 +1829,48 @@ public static boolean fromTheSameCaseExpression(@NotNull PsiElement origin, @Not
...
+      ErlangFunction function = resolveToFunction(call);
+      if (function == null) return false;
+
+      int position = getPositionOfArgument(argument);
+      if (position == -1) return false;
+
+      for (ErlangFunctionClause clause : function.getFunctionClauseList()) {
+        List<ErlangArgumentDefinition> definitions = clause.getArgumentDefinitionList().getArgumentDefinitionList();
+        ErlangArgumentDefinition definition = definitions.size() > position ? definitions.get(position) : null;
+        ErlangQVar var = PsiTreeUtil.findChildOfType(definition, ErlangQVar.class);
```
Comment (deadok22) [inline src/org/intellij/erlang/psi/impl/ErlangPsiImplUtil.java]: Again, you're searching recursively - that's not what you want to do.

File: src/org/intellij/erlang/psi/impl/ErlangPsiImplUtil.java [hunk 3]
```diff
@@ -1821,6 +1829,48 @@ public static boolean fromTheSameCaseExpression(@NotNull PsiElement origin, @Not
...
+
+      return false;
+    }
+
+    private static int getPositionOfArgument(@NotNull ErlangQAtom argument) {
+      ErlangArgumentList argumentList = PsiTreeUtil.getParentOfType(argument, ErlangArgumentList.class);
+      ErlangExpression expression = PsiTreeUtil.getParentOfType(argument, ErlangExpression.class, false);
+      if (expression != null && expression.getParent() instanceof ErlangColonQualifiedExpression) {
+        expression = (ErlangExpression) expression.getParent();
+      }
```
Comment (deadok22) [inline src/org/intellij/erlang/psi/impl/ErlangPsiImplUtil.java]: ObjectUtils.tryCast(expression != null ? expression.getParent() : null, ErlangColonQualifiedExpression.class);

**Comment by deadok22** [comment]
You can also use function specs to find out types of your arguments.

File: src/org/intellij/erlang/ErlangReferenceContributor.java [hunk 4]
```diff
@@ -29,30 +31,39 @@
...
+    registerRecords(registrar, "erlang", "is_record", 2, 1);
+    registerRecords(registrar, "erlang", "is_record", 3, 1);
+
+    ErlangFunctionCallNamedParameter<PsiElement> moduleParameter =
+      new ErlangFunctionCallNamedParameter<PsiElement>("Module", "Mod", "_Module");
+    registrar.registerReferenceProvider(psiElement(ErlangQAtom.class).with(moduleParameter), getProvider(Kind.MODULE));
+
+    ErlangFunctionCallNamedParameter<PsiElement> funParameter =
+      new ErlangFunctionCallNamedParameter<PsiElement>("Function", "Fun", "_Function");
+    registrar.registerReferenceProvider(psiElement(ErlangQAtom.class).with(funParameter), getProvider(Kind.FUNCTION));
```
Comment (deadok22) [inline src/org/intellij/erlang/ErlangReferenceContributor.java]: It won't work really well, as you will only create a reference to function name, which will only resolve to local function declarations - you can do something better here, I believe.

File: src/org/intellij/erlang/psi/impl/ErlangPsiImplUtil.java (hunk 1 above)
Comment (telezhnaya) [inline src/org/intellij/erlang/psi/impl/ErlangPsiImplUtil.java]: `com.intellij.psi.util.PsiTreeUtil#getParentOfType(com.intellij.psi.PsiElement, java.lang.Class<T>)` is not recursive. It uses while loop.

Could you suggest another way to get function call expression, without loops and recursion?
Or maybe you mean something else?

Comment (deadok22) [inline src/org/intellij/erlang/psi/impl/ErlangPsiImplUtil.java]: I mean it's recursive in a way it works - it loops through all of the parents, whereas you only need to check like a parent, or, may be a grandparent.

Take a look at how it's done in another pattern which was used before your changes.

Just so you comprehend the problem here: ```foo({1, {2, .....{N, atom}}}}}})``` will make you traverse all the way up through N levels of nested tuples to get to the function call expression.

File: src/org/intellij/erlang/psi/impl/ErlangPsiImplUtil.java (hunk 3 above)
Comment (telezhnaya) [inline src/org/intellij/erlang/psi/impl/ErlangPsiImplUtil.java]: Let me explain. When we pass a function as parameter, there are 2 scenarios: `my_fun(foo)` or `my_fun(main:foo)`.
So I don't need to cast it to ErlangColonQualifiedExpression. I just need to get one level up, if it's colon qualified. If not - I need to do nothing.

Comment (deadok22) [inline src/org/intellij/erlang/psi/impl/ErlangPsiImplUtil.java]: AFAIK, main:foo is not a valid expression.

Comment (telezhnaya) [inline src/org/intellij/erlang/psi/impl/ErlangPsiImplUtil.java]: I tested it. It works good.

Comment (deadok22) [inline src/org/intellij/erlang/psi/impl/ErlangPsiImplUtil.java]: What do you mean by 'works good'? It just won't compile.

**Comment by telezhnaya** [comment]
Have a look, please.

**Comment by deadok22** [comment]
You don't need these 'incl' modules in your tests. Please, clean it up.

Also, I noticed, you haven't improved function references - why have them at all, then? The way it's done it will only lead to incorrect resolution of functions referenced in calls with parameters of module, function, args.

**Pull request closed without merging**